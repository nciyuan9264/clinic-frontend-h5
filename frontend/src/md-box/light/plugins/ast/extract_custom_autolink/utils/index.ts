export { HyperNodeType, getHyperLinkNodes } from './parser';

export const completeProtocol = (raw: string) => {
  if (raw.match(/https?:\/\//)) {
    return raw;
  }
  return `https://${raw}`;
};
