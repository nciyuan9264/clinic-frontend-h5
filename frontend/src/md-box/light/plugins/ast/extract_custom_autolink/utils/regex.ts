// -> refer @ea/url-recognition -> refer 飞书云文档
const PROTOCOL = 'https?|s?ftp|ftps|nfs|ssh';
const LINK_SUFFIX =
  "\\b([/#?][\\-A-Z0-9@:%_+.~#?&'/=;$,!\\*\\[\\]\\(\\){}^|<>]*[-A-Z0-9&@#/%=~\\(\\)_|]?)?";
const LOCALHOST_REG = /localhost(:[0-9]{2,5})?/;
const IP_ADDRESS_REG =
  // eslint-disable-next-line max-len
  /(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}(2[0-4][0-9]|25[0-5]|1[0-9]{2}|[1-9][0-9]|[0-9])(:[0-9]{2,5})?/;
const IP_REG = new RegExp(
  `((((${PROTOCOL})://)?(${LOCALHOST_REG.source}))|(((${PROTOCOL})://)(${IP_ADDRESS_REG.source})))${LINK_SUFFIX}`,
  'gi',
);
const URL_HOST_BODY = '([-A-Z0-9%_+~#@][\\-A-Z:0-9%_+~#@]{0,256}\\.){1,50}';
const ANY_TOP_DOMAIN = '[a-z\\-]{2,15}';
const TOP_DOMAIN =
  'com|org|net|int|edu|gov|mil|ac|ad|ae|af|ag|ai|al|am|an|ao|aq|ar|as|at|au|aw|ax|az|ba|bb|bd|be|bf|bg|bh|bi|bj|bm|bn|bo|br|bs|bt|bv|bw|by|bz|ca|cc|cf|cg|ch|ci|ck|cl|cm|cn|co|cr|cu|cv|cw|cx|cy|cz|de|dj|dk|dm|do|dz|ec|ee|eg|eh|er|es|et|eu|fi|fj|fk|fm|fo|fr|ga|gb|gd|ge|gf|gh|gi|gl|gm|gn|gp|gr|gs|gt|gu|gw|gy|hk|hm|hn|hr|ht|hu|id|ie|il|im|in|io|iq|ir|is|it|jm|jo|jp|ke|kg|kh|ki|km|kn|kp|kr|kw|ky|kz|la|lb|lc|li|lk|lr|ls|lt|lu|lv|ly|ma|mc|md|me|mf|mg|mh|mk|ml|mm|mn|mo|mp|mq|mr|ms|mt|mu|mv|mw|mx|my|mz|na|nc|ne|nf|ng|ni|nl|no|np|nr|nu|nz|om|pa|pe|pf|pg|ph|pk|pl|pm|pn|pr|ps|pt|pw|py|qa|re|ro|rs|ru|rw|sa|sb|sc|sd|se|sg|sh|si|sj|sk|sl|sm|sn|so|sr|ss|st|su|sv|sx|sy|sz|tc|td|tf|tg|th|tj|tk|tl|tm|tn|to|tp|tr|tt|tv|tw|tz|ua|ug|uk|um|us|uy|uz|va|vc|ve|vg|vi|vn|vu|wf|ws|ye|yt|za|zm|zw|site|top|wtf|xxx|xyz|cloud|engineering|help|one';
const TOP_DOMAIN_LITE = 'com|cn|tk|de|net|org|uk|info|nl|ru|hk';
// 由于顶级域名不全，故更改为带 协议头的url 忽略上面顶级域名，直接匹配
const URL_REG = new RegExp(
  // eslint-disable-next-line max-len
  `(((${PROTOCOL}):\\/\\/${URL_HOST_BODY}${ANY_TOP_DOMAIN})|(${URL_HOST_BODY}(${TOP_DOMAIN_LITE})))(:[0-9]{2,5})?${LINK_SUFFIX}`,
  'gi',
);
// mailto的贪婪匹配在遇到极长的字符串（无空格的英文）会很慢，需要限长。
const EMAIL_URL_BODY = "[\\w.!#$%&'*+-/=?^_\\`{|}~]{1,2000}@[A-Za-z0-9_.-]+\\.";
const LENGTH_LIMITED_EMAIL_REG = new RegExp(
  `^(?!data:)((mailto:${EMAIL_URL_BODY}${ANY_TOP_DOMAIN})|(${EMAIL_URL_BODY}(${TOP_DOMAIN})))\\b`,
  'gi',
);

export const urlRegExp = new RegExp(
  `(${IP_REG.source})|(${URL_REG.source})|(${LENGTH_LIMITED_EMAIL_REG.source})`,
  'gi',
);
