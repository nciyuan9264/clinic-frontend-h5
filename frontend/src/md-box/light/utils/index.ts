export * from './micromark';
export * from './marked';
export * from './renderer';
export * from './common';
export * from './hop';
export * from './plugin';
