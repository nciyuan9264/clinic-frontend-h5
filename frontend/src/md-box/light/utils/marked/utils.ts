/** 源自 markedjs 源码 */

const escapeTest = /[&<>"']/;
const escapeReplace = new RegExp(escapeTest.source, 'g');
const escapeTestNoEncode = /[<>"']|&(?!(#\d{1,7}|#[Xx][a-fA-F0-9]{1,6}|\w+);)/;
const escapeReplaceNoEncode = new RegExp(escapeTestNoEncode.source, 'g');

const escapeReplacements: Record<string, string> = {
  '&': '&amp;',
  '<': '&lt;',
  '>': '&gt;',
  '"': '&quot;',
  "'": '&#39;',
};

export const escapeMarkdownText = (html: string, encode = false) => {
  const getEscapeReplacement = (ch: string) => escapeReplacements[ch];

  if (encode) {
    if (escapeTest.test(html)) {
      return html.replace(escapeReplace, getEscapeReplacement);
    }
  } else {
    if (escapeTestNoEncode.test(html)) {
      return html.replace(escapeReplaceNoEncode, getEscapeReplacement);
    }
  }
  return html;
};
