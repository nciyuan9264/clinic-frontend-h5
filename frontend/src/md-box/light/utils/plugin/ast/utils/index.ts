export * from './update_ast';
