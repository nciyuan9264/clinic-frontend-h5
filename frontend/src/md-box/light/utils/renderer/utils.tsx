import React, { createElement } from 'react';

import { isRegExp } from 'lodash-es';
import { DOMNode, Element, Text, domToReact } from 'html-react-parser';

import { RenderFunction } from '../../type';

export const isTextType = (node: DOMNode): node is Text => {
  return node.type === 'text';
};

export const isEndlineText = (node: DOMNode): node is Text => {
  return isTextType(node) && /^\n+$/.test(getInnerText(node));
};

/**
 * 此处断言的是Element的子类型：即为A标签的类型，但是Element不可进一步进行类型收窄，因此
 * 返回node is Element
 */
export const isElementType = (
  node: DOMNode,
  ...names: (string | RegExp)[]
): node is Element => {
  if (!('attribs' in node) || typeof node.attribs !== 'object') {
    return false;
  }
  return (
    node.type === 'tag' &&
    (!names?.length ||
      names.some((item) => {
        if (isRegExp(item)) {
          return item.test(node.name);
        } else {
          return item === node.name;
        }
      }))
  );
};

export const getInnerText = (node: DOMNode): string => {
  if (isTextType(node)) {
    return node.data;
  }
  if (isElementType(node)) {
    return node.children.map(getInnerText).join('');
  }
  return '';
};

/**
 * 检测字符或字符串是否包含中文字符
 * https://github.com/anton-bot/contains-chinese/blob/master/index.js#L7
 * @param char 字符或字符串
 * @returns 是否包含中文字符
 */
export const detectIsChinese = (char: string) => {
  const HAN_REGEX =
    // eslint-disable-next-line max-len
    /[\u2E80-\u2E99\u2E9B-\u2EF3\u2F00-\u2FD5\u3005\u3007\u3021-\u3029\u3038-\u303B\u3400-\u4DB5\u4E00-\u9FD5\uF900-\uFA6D\uFA70-\uFAD9]/;
  return HAN_REGEX.test(char);
};

export const renderDom: RenderFunction<false> = (node) => {
  return <>{domToReact([node])}</>;
};

export const renderReactElement = <P extends Record<string, unknown>>(
  tag: string,
  props: P,
  children?: React.ReactNode[],
) => {
  if (props.className === 'markdown-root') {
    return <>{children}</>;
  } else {
    return (
      <>{createElement(tag, props, children?.length ? children : undefined)}</>
    );
  }
};
