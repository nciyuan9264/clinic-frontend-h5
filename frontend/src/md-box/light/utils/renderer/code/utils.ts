import { DOMNode, Element } from 'html-react-parser';

import { isElementType } from '../utils';

const DEFAULT_LANGUAGE = 'plaintext';

const getCodeElement = (node: DOMNode): Element | undefined => {
  if (isElementType(node, 'code')) {
    return node;
  }

  if (isElementType(node, 'pre')) {
    return node.childNodes.find((childNode): childNode is Element =>
      isElementType(childNode, 'code'),
    );
  }
};

export const getLanguageFromElement = (
  node: DOMNode,
  defaultLanguage = DEFAULT_LANGUAGE,
): string => {
  const codeElement = getCodeElement(node);

  if (codeElement) {
    return (
      codeElement.attribs?.class?.match(/language-(.*)/)?.[1] || defaultLanguage
    );
  }

  return defaultLanguage;
};

export const getMetaFromElement = (node: DOMNode, removeBuiltIn = true) => {
  const codeElement = getCodeElement(node);

  const metaString = codeElement?.attribs['data-meta'];

  if (!metaString) {
    return;
  }

  if (!removeBuiltIn) {
    return metaString;
  }

  return (
    metaString
      .split(/\s+/)
      .filter((item) => !item.match(/^__[^_].*/))
      .join(' ') || undefined
  );
};

export const getIndentFromCodeElememt = (node: DOMNode) => {
  const metaString = getMetaFromElement(node, false);

  if (!metaString) {
    return 0;
  }

  const matchResult = metaString.match(
    /(\s|^)__indent=(?<indent>[0-9]+)(\s|$)/,
  );

  if (!matchResult) {
    return 0;
  }

  const indentValue = Number(matchResult.groups?.indent);

  return isNaN(indentValue) ? 0 : indentValue;
};
