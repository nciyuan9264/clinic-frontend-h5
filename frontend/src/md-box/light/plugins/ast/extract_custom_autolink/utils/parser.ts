import emailValidator from 'email-validator';

import { urlRegExp } from './regex';

export enum HyperNodeType {
  text,
  link,
}

export type HyperNode =
  | {
      type: HyperNodeType.text;
      text: string;
    }
  | {
      type: HyperNodeType.link;
      url: string;
    };

export type HyperNodeTitled = HyperNode & {
  title: string;
};

const openAndCloseCharPairs: Record<string, string> = {
  '(': ')',
  '[': ']',
  '{': '}',
  '<': '>',
};
const openChars = Object.keys(openAndCloseCharPairs);
const closeChars = Object.values(openAndCloseCharPairs);
const unreasonableSuffix = ['.', ',', '!', '?', "'", ':', ';', '@', '#'];

/**
 * 分割合理的URL
 * @returns 返回字符串中合理的URL的长度，范围[0, input.length]
 */
const getAReasonableUrl = (input: string) => {
  let index = -1;
  const stack: Array<{
    char: string;
    index: number;
  }> = [];
  for (let i = 0; i < input.length; ++i) {
    const char = input[i];
    if (openChars.includes(char)) {
      stack.push({
        char,
        index: i,
      });
    } else if (closeChars.includes(char)) {
      if (!stack.length) {
        index = i;
        break;
      } else {
        const { char: lastOpenChar } = stack[stack.length - 1];
        if (char === openAndCloseCharPairs[lastOpenChar]) {
          stack.pop();
        } else {
          ({ index } = stack[0]);
          break;
        }
      }
    }
  }
  if (stack.length) {
    ({ index } = stack[0]);
  }
  if (index === -1) {
    index = input.length;
  }

  while (index > 1 && unreasonableSuffix.includes(input[index - 1])) {
    index--;
  }
  return input.slice(0, index);
};

/**
 * 分割字符串中的超链接和常规字符，输出格式化数据 HyperNode[]
 */
export const getHyperLinkNodes = (str: string): HyperNode[] => {
  const nodes: HyperNode[] = [];
  let cur = 0;
  urlRegExp.lastIndex = 0;
  while (true) {
    const result = urlRegExp.exec(str);
    if (!result) {
      break;
    }

    const { index, 0: content } = result;
    const text = str.slice(cur, index);

    // ① 文字部分
    if (text) {
      nodes.push({
        type: HyperNodeType.text,
        text,
      });
    }

    // ② 链接/markdown/email 部分，email 被渲染为文字，code block / inline code 中的链接也当做文字
    if (content) {
      /** 1. 非正常前缀 */
      const abnormalPrefixMatchResult = /^(@|#|\+|-|%)+/.exec(content);
      if (abnormalPrefixMatchResult) {
        nodes.push({
          type: HyperNodeType.text,
          text: abnormalPrefixMatchResult[0],
        });
        cur = result.index + abnormalPrefixMatchResult[0].length;
        urlRegExp.lastIndex = cur;
      } else {
        /** 2. 合理Url作为前缀 */
        const subUrl = getAReasonableUrl(content);

        if (subUrl) {
          if (emailValidator.validate(subUrl)) {
            nodes.push({
              type: HyperNodeType.text,
              text: subUrl,
            });
          } else {
            nodes.push({
              type: HyperNodeType.link,
              url: subUrl,
            });
          }

          cur = result.index + subUrl.length;
          urlRegExp.lastIndex = cur;
        } else {
          const skipStr = content.slice(0, 1);

          nodes.push({
            type: HyperNodeType.text,
            text: skipStr,
          });

          cur = result.index + skipStr.length;
          urlRegExp.lastIndex = cur;
        }
      }
    }
  }
  if (cur !== str.length) {
    nodes.push({ type: HyperNodeType.text, text: str.slice(cur) });
  }
  return nodes;
};
