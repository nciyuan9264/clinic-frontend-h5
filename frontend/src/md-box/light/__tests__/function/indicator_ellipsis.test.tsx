import { FC } from 'react';

import { render, act } from '@testing-library/react';

import { MdBoxForTesting } from '../utils';
import { MdBoxIndicatorProps } from '../../../contexts';

// eslint-disable-next-line max-lines-per-function
describe('打字机指示点测试', () => {
  test('文本后展示打字机指示点', async () => {
    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={`
第一行
`}
          showIndicator
        />,
      ),
    );

    expect(rootView.asFragment()).toMatchSnapshot();

    expect(rootView.queryByTestId('indicator')).toBeInTheDocument();
  });

  test('链接后展示打字机指示点', async () => {
    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={`
[link](http://bytedance.com)
`}
          showIndicator
        />,
      ),
    );

    expect(rootView.asFragment()).toMatchSnapshot();

    expect(rootView.queryByTestId('indicator')).toBeInTheDocument();
  });

  test('表格内展示打字机指示点', async () => {
    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={`
| foo | bar |
| --- | --- |
| baz | bim
`}
          showIndicator
        />,
      ),
    );

    expect(rootView.asFragment()).toMatchSnapshot();

    expect(rootView.queryByTestId('indicator')).toBeInTheDocument();
  });

  test('图片后不展示打字机指示点', async () => {
    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={`
![link](http://bytedance.com/logo.png)
`}
          showIndicator
        />,
      ),
    );

    expect(rootView.asFragment()).toMatchSnapshot();

    expect(rootView.queryByTestId('indicator')).not.toBeInTheDocument();
  });

  test('代码块后不展示打字机指示点', async () => {
    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={`
\`\`\`
print("hello")
\`\`\`
`}
          showIndicator
        />,
      ),
    );

    expect(rootView.asFragment()).toMatchSnapshot();

    expect(rootView.queryByTestId('indicator')).not.toBeInTheDocument();
  });

  test('打字机指示点与省略号共存', async () => {
    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={`
  第一行
  `}
          showIndicator
          showEllipsis
        />,
      ),
    );

    expect(rootView.asFragment()).toMatchSnapshot();

    expect(rootView.queryByTestId('indicator')).toBeInTheDocument();

    expect(rootView.baseElement.textContent?.trim()).toBe('第一行...');
  });

  test('二次渲染', async () => {
    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={`
第一行
`}
          showIndicator
        />,
      ),
    );

    const prevFragment = rootView.asFragment();

    await act(() => {
      rootView.rerender(
        <MdBoxForTesting
          markDown={`
第一行
`}
        />,
      );
    });

    expect(prevFragment).toMatchDiffSnapshot(rootView.asFragment());
  });

  test('自定义指示点', async () => {
    const CustomIndicator: FC<MdBoxIndicatorProps> = () => {
      return (
        <span style={{ color: 'grey' }} data-testid="custom-indicator">
          Waiting
        </span>
      );
    };

    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={`
第一行
`}
          showIndicator
          slots={{
            Indicator: CustomIndicator,
          }}
        />,
      ),
    );

    expect(rootView.asFragment()).toMatchSnapshot();

    expect(rootView.queryByTestId('custom-indicator')).toBeInTheDocument();
  });
});

describe('省略号插入测试', () => {
  test('文本后插入省略号', async () => {
    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={`
第一行
`}
          showEllipsis
        />,
      ),
    );

    expect(rootView.asFragment()).toMatchSnapshot();

    expect(rootView.baseElement.textContent?.trim()).toBe('第一行...');
  });

  test('行内内容后插入省略号', async () => {
    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={`
\`print("hello")\`
`}
          showEllipsis
        />,
      ),
    );

    expect(rootView.asFragment()).toMatchSnapshot();

    expect(rootView.baseElement.textContent?.trim()).toBe('print("hello")...');
  });

  test('图片后不插入省略号', async () => {
    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={`
prefix![link](http://bytedance.com/logo.png)
`}
          showEllipsis
        />,
      ),
    );

    expect(rootView.asFragment()).toMatchSnapshot();

    expect(rootView.baseElement.textContent?.trim()).toBe('prefix');
  });

  test('二次渲染', async () => {
    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={`
第一行
`}
          showEllipsis
        />,
      ),
    );

    const prevFragment = rootView.asFragment();

    await act(() => {
      rootView.rerender(
        <MdBoxForTesting
          markDown={`
第一行
`}
        />,
      );
    });

    expect(prevFragment).toMatchDiffSnapshot(rootView.asFragment());
  });
});
