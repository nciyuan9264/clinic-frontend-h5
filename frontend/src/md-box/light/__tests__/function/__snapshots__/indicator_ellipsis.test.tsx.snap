// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`打字机指示点测试 二次渲染 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,24 +1,16 @@
  <DocumentFragment>
    <div
      class="container flow-markdown-body theme-default"
-     data-show-indicator="true"
+     data-show-indicator="false"
      dir="ltr"
      theme-mode="light"
    >
      <div
        class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
      >
        第一行
-       <span
-         class="container"
-       >
-         <span
-           class="indicator flashing"
-           data-testid="indicator"
-         />
-       </span>
      </div>
      <br
        class="container wrapper non-select"
      />
      "
`;

exports[`打字机指示点测试 代码块后不展示打字机指示点 1`] = `
<DocumentFragment>
  <div
    class="container flow-markdown-body theme-default"
    data-show-indicator="true"
    dir="ltr"
    theme-mode="light"
  >
    <div
      class="container code-group-item--normal hide-indicator"
      data-testid="code_block"
    >
      <div
        class="code-area"
        dir="ltr"
      >
        <div
          class="header"
        >
          <div
            class="text"
          >
            <div
              class="icon"
            />
            plaintext
          </div>
          <div
            class="actions"
          >
            <div
              class="item"
            >
              <div
                class="icon hoverable"
                data-testid="code_block_copy"
              >
                <div
                  class="img"
                />
              </div>
            </div>
            <div
              class="item"
            >
              <div
                class="icon hoverable"
                data-testid="code_block_mode_switch"
              >
                <div
                  class="img"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="content content--plaintext"
        >
          <pre
            class="container language-plaintext"
          >
            <code
              class="language-plaintext"
            >
              print("hello")

            </code>
          </pre>
        </div>
      </div>
    </div>
    <br
      class="container wrapper non-select"
    />
  </div>
</DocumentFragment>
`;

exports[`打字机指示点测试 图片后不展示打字机指示点 1`] = `
<DocumentFragment>
  <div
    class="container flow-markdown-body theme-default"
    data-show-indicator="true"
    dir="ltr"
    theme-mode="light"
  >
    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      <div
        class="container"
      >
        <div
          class="image-wrapper"
          data-testid="mdbox_image"
          style="width: 100%; height: 0px; max-width: 400px;"
        >
          <span
            style="display: inline-block;"
          >
            <img
              alt="image"
              class="image"
              decoding="async"
              height="256"
              imagex-type="react"
              imagex-version="0.3.7"
              loading="lazy"
              src="http://bytedance.com/logo.png"
              style="border-radius: 12px; overflow: hidden; filter: blur(20px); background-size: cover; background-image: url(data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTAwJSIgaGVpZ2h0PSIxMDAlIiB2aWV3Qm94PSIwLDAsMjAlLDEwMCUiIHZlcnNpb249IjEuMSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB4bWxuczp4bGluaz0iaHR0cDovL3d3dy53My5vcmcvMTk5OS94bGluayI+CiAgPGRlZnM+CiAgICA8bGluZWFyR3JhZGllbnQgaWQ9ImciPgogICAgICA8c3RvcCBzdG9wLWNvbG9yPSIjRjJGM0Y1IiBvZmZzZXQ9IjI1JSIgLz4KICAgICAgPHN0b3Agc3RvcC1jb2xvcj0iI0U1RTZFQiIgb2Zmc2V0PSIzNyUiIC8+CiAgICAgIDxzdG9wIHN0b3AtY29sb3I9IiNGMkYzRjUiIG9mZnNldD0iNjMlIiAvPgogICAgPC9saW5lYXJHcmFkaWVudD4KICA8L2RlZnM+CiAgPHJlY3QgaWQ9InIiIHdpZHRoPSI0MDAlIiBoZWlnaHQ9IjEwMCUiIGZpbGw9InVybCgjZykiIC8+CiAgPGFuaW1hdGUgeGxpbms6aHJlZj0iI3IiIGF0dHJpYnV0ZU5hbWU9IngiIGZyb209Ii0zMDAlIiB0bz0iMCUiIGR1cj0iMS41cyIgcmVwZWF0Q291bnQ9ImluZGVmaW5pdGUiICAvPgo8L3N2Zz4=); background-position: center;"
              width="400"
            />
          </span>
        </div>
      </div>
    </div>
    <br
      class="container wrapper non-select"
    />
    

  </div>
</DocumentFragment>
`;

exports[`打字机指示点测试 打字机指示点与省略号共存 1`] = `
<DocumentFragment>
  <div
    class="container flow-markdown-body theme-default"
    data-show-indicator="true"
    dir="ltr"
    theme-mode="light"
  >
    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      第一行...
      <span
        class="container"
      >
        <span
          class="indicator flashing"
          data-testid="indicator"
        />
      </span>
    </div>
    <br
      class="container wrapper non-select"
    />
    

  </div>
</DocumentFragment>
`;

exports[`打字机指示点测试 文本后展示打字机指示点 1`] = `
<DocumentFragment>
  <div
    class="container flow-markdown-body theme-default"
    data-show-indicator="true"
    dir="ltr"
    theme-mode="light"
  >
    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      第一行
      <span
        class="container"
      >
        <span
          class="indicator flashing"
          data-testid="indicator"
        />
      </span>
    </div>
    <br
      class="container wrapper non-select"
    />
    

  </div>
</DocumentFragment>
`;

exports[`打字机指示点测试 自定义指示点 1`] = `
<DocumentFragment>
  <div
    class="container flow-markdown-body theme-default"
    data-show-indicator="true"
    dir="ltr"
    theme-mode="light"
  >
    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      第一行
      <span
        data-testid="custom-indicator"
        style="color: grey;"
      >
        Waiting
      </span>
    </div>
    <br
      class="container wrapper non-select"
    />
    

  </div>
</DocumentFragment>
`;

exports[`打字机指示点测试 表格内展示打字机指示点 1`] = `
<DocumentFragment>
  <div
    class="container flow-markdown-body theme-default"
    data-show-indicator="true"
    dir="ltr"
    theme-mode="light"
  >
    <div
      class="auto-hide-last-sibling-br table-container"
    >
      <table>
        <thead>
          <tr>
            <th>
              foo
            </th>
            <th>
              bar
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              baz
            </td>
            <td>
              bim
              <span
                class="container"
              >
                <span
                  class="indicator flashing"
                  data-testid="indicator"
                />
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <br
      class="container wrapper non-select"
    />
    

  </div>
</DocumentFragment>
`;

exports[`打字机指示点测试 链接后展示打字机指示点 1`] = `
<DocumentFragment>
  <div
    class="container flow-markdown-body theme-default"
    data-show-indicator="true"
    dir="ltr"
    theme-mode="light"
  >
    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      <a
        class="link"
        href="http://bytedance.com"
        target="_blank"
      >
        link
        <span
          class="container"
        >
          <span
            class="indicator flashing"
            data-testid="indicator"
          />
        </span>
      </a>
    </div>
    <br
      class="container wrapper non-select"
    />
    

  </div>
</DocumentFragment>
`;

exports[`省略号插入测试 二次渲染 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -6,11 +6,11 @@
      theme-mode="light"
    >
      <div
        class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
      >
-       第一行...
+       第一行
      </div>
      <br
        class="container wrapper non-select"
      />
      "
`;

exports[`省略号插入测试 图片后不插入省略号 1`] = `
<DocumentFragment>
  <div
    class="container flow-markdown-body theme-default"
    data-show-indicator="false"
    dir="ltr"
    theme-mode="light"
  >
    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      prefix
      <div
        class="container"
        style="margin-top: 12px;"
      >
        <div
          class="image-wrapper"
          data-testid="mdbox_image"
          style="width: 100%; height: 0px; max-width: 400px;"
        >
          <span
            style="display: inline-block;"
          >
            <img
              alt="image"
              class="image"
              decoding="async"
              height="256"
              imagex-type="react"
              imagex-version="0.3.7"
              loading="lazy"
              src="http://bytedance.com/logo.png"
              style="border-radius: 12px; overflow: hidden; filter: blur(20px); background-size: cover; background-image: url(data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTAwJSIgaGVpZ2h0PSIxMDAlIiB2aWV3Qm94PSIwLDAsMjAlLDEwMCUiIHZlcnNpb249IjEuMSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB4bWxuczp4bGluaz0iaHR0cDovL3d3dy53My5vcmcvMTk5OS94bGluayI+CiAgPGRlZnM+CiAgICA8bGluZWFyR3JhZGllbnQgaWQ9ImciPgogICAgICA8c3RvcCBzdG9wLWNvbG9yPSIjRjJGM0Y1IiBvZmZzZXQ9IjI1JSIgLz4KICAgICAgPHN0b3Agc3RvcC1jb2xvcj0iI0U1RTZFQiIgb2Zmc2V0PSIzNyUiIC8+CiAgICAgIDxzdG9wIHN0b3AtY29sb3I9IiNGMkYzRjUiIG9mZnNldD0iNjMlIiAvPgogICAgPC9saW5lYXJHcmFkaWVudD4KICA8L2RlZnM+CiAgPHJlY3QgaWQ9InIiIHdpZHRoPSI0MDAlIiBoZWlnaHQ9IjEwMCUiIGZpbGw9InVybCgjZykiIC8+CiAgPGFuaW1hdGUgeGxpbms6aHJlZj0iI3IiIGF0dHJpYnV0ZU5hbWU9IngiIGZyb209Ii0zMDAlIiB0bz0iMCUiIGR1cj0iMS41cyIgcmVwZWF0Q291bnQ9ImluZGVmaW5pdGUiICAvPgo8L3N2Zz4=); background-position: center;"
              width="400"
            />
          </span>
        </div>
      </div>
    </div>
    <br
      class="container wrapper non-select"
    />
    

  </div>
</DocumentFragment>
`;

exports[`省略号插入测试 文本后插入省略号 1`] = `
<DocumentFragment>
  <div
    class="container flow-markdown-body theme-default"
    data-show-indicator="false"
    dir="ltr"
    theme-mode="light"
  >
    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      第一行...
    </div>
    <br
      class="container wrapper non-select"
    />
    

  </div>
</DocumentFragment>
`;

exports[`省略号插入测试 行内内容后插入省略号 1`] = `
<DocumentFragment>
  <div
    class="container flow-markdown-body theme-default"
    data-show-indicator="false"
    dir="ltr"
    theme-mode="light"
  >
    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      <code>
        print("hello")
      </code>
      ...
    </div>
    <br
      class="container wrapper non-select"
    />
    

  </div>
</DocumentFragment>
`;
